### ISSUE TITLE
monitor_gradients=True leads to `∂None`

### ISSUE THREAD

**Issue by rodrigob** [lmjohns3/downhill#8]
The downhill log includes
`I 2015-09-03 23:19:54 downhill.base:231 Adam 1 loss=0.613750 ∂None=0.001453`

what should I do make the `∂None` become something with a sensible name ?

**Comment by lmjohns3** [comment]
Hm, I think you need to provide a name for the parameter variables in your model. For example,

    w = theano.shared(np.random.randn(3, 4), name='foo')

The ``'foo'`` name will be used in the gradient monitor.

**Comment by rodrigob** [comment]
I am currently optimizing a http://keras.io model, there the parameters is a list of shared variables (and thus nameless).

**Comment by lmjohns3** [comment]
Hm, that's unfortunate. Can you go through the list of parameters and assign names to them manually? Downhill doesn't have any way of knowing what the names for parameters in a generic loss should be.

**Comment by rodrigob** [comment]
they do have individual names (of the type `auto_12` if not set by the user); but maybe downhill does not handle such lists ?

**Comment by rodrigob** [comment]
ok my bad. When adding names I get to see what is expected.
However when the names are None I only get one "grad(None)"; while with non-None I see "grad(NameA) grad(NameB) grad(NameC)". I would have expected to see then "grad(None) grad(None) grad(None)".

**Comment by lmjohns3** [comment]
Good, I'm glad you're seeing something useful after providing names.

About the "grad(None)" cardinality -- this is happening because the monitors are stored in a dictionary internally (and returned to the caller from ``minimize`` as a dictionary as well), so there is a key collision when multiple monitors have the same name.

I could work around this internally by manually assigning distinct parameter names for parameters that do not have them. The names will be meaningless ("param1", "param2", etc.) but at least they won't collide.

It's still definitely a better idea to name the parameters in your model if possible.

**Comment by rodrigob** [comment]
Ahh I understand, that makes sense.
Yes I agree it matters most to have proper names for the parameters.
I wonder if printing "grad(None)" is very user friendly.
I would rather go for generating a warning once (since downhill uses a logging system anyways), and then being silent.

As new user I though that "grad(None)" was the average norm of the different parameters, while now I see it was just "a norm" (probably last element).
